core/vm/instructions.go

```diff
@@ -0,0 +1,588 @@
+// Copyright 2014 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package vm
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+type instrFn func(instr instruction, env Environment, context *Context, memory *Memory, stack *stack)
+type instrExFn func(instr instruction, ret *big.Int, env Environment, context *Context, memory *Memory, stack *stack)
+
+type instruction struct {
+	op     OpCode
+	pc     uint64
+	fn     instrFn
+	specFn instrExFn
+	data   *big.Int
+
+	gas   *big.Int
+	spop  int
+	spush int
+}
+
+func opStaticJump(instr instruction, ret *big.Int, env Environment, context *Context, memory *Memory, stack *stack) {
+	ret.Set(instr.data)
+}
+
+func opAdd(instr instruction, env Environment, context *Context, memory *Memory, stack *stack) {
+	x, y := stack.pop(), stack.pop()
+
+	stack.push(U256(new(big.Int).Add(x, y)))
+}
+
+func opSub(instr instruction, env Environment, context *Context, memory *Memory, stack *stack) {
+	x, y := stack.pop(), stack.pop()
+
+	stack.push(U256(new(big.Int).Sub(x, y)))
+}
+
+func opMul(instr instruction, env Environment, context *Context, memory *Memory, stack *stack) {
+	x, y := stack.pop(), stack.pop()
+
+	stack.push(U256(new(big.Int).Mul(x, y)))
+}
+
+func opDiv(instr instruction, env Environment, context *Context, memory *Memory, stack *stack) {
+	base := new(big.Int)
+	x, y := stack.pop(), stack.pop()
+
+	if y.Cmp(common.Big0) != 0 {
+		base.Div(x, y)
+	}
+
+	// pop result back on the stack
+	stack.push(U256(base))
+}
+
+func opSdiv(instr instruction, env Environment, context *Context, memory *Memory, stack *stack) {
+	base := new(big.Int)
+	x, y := S256(stack.pop()), S256(stack.pop())
+
+	if y.Cmp(common.Big0) == 0 {
+		base.Set(common.Big0)
+	} else {
+		n := new(big.Int)
+		if new(big.Int).Mul(x, y).Cmp(common.Big0) < 0 {
+			n.SetInt64(-1)
+		} else {
+			n.SetInt64(1)
+		}
+
+		base.Div(x.Abs(x), y.Abs(y)).Mul(base, n)
+
+		U256(base)
+	}
+
+	stack.push(base)
```

Comment:
Sorry, that was a bit misguided. We also push `common.BigFalse` and friends to the stack in other places.
Bigints on the stack are not mutated. I think what @Gustav-Simonsson meant is that we could do something
like this to avoid allocating `base` when it is not required.

```go
x, y := S256(stack.pop()), S256(stack.pop())
if y.Cmp(common.Big0) == 0 {
    stack.Push(common.Big0)
    return
}
...
```